update data_container example to pymc 5.6 (#559)
pymc-devs/pymc-examples: examples/howto/data_container.myst.md

---

(data_container)=
# Using Data Containers

:::{post} Dec 16, 2021
:tags: posterior predictive, shared data 
:category: beginner
:author: Juan Martin Loyola, Kavya Jaiswal, Oriol Abril, Jesse Grabowski
:::

```{code-cell} ipython3
import arviz as az
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import pymc as pm

from numpy.random import default_rng

print(f"Running on PyMC v{pm.__version__}")
```

```{code-cell} ipython3
%matplotlib inline
%config InlineBackend.figure_format = 'retina'

> RANDOM_SEED = sum(map(ord, "Data Containers in PyMC5"))

Reviewer: ```suggestion
RANDOM_SEED = sum(map(ord, "Data Containers in PyMC"))
```

rng = default_rng(RANDOM_SEED)
az.style.use("arviz-darkgrid")
```

## Data Containers

> The {class}`pymc.Data` container allows users to keep track of dimensions (like dates or cities) and coordinates (such as the actual date times or city names) of multi-dimensional data. It offers two wrappers for this: {class}`pymc.ConstantData` and {class}`MutableData`. Both allow you to specify the dimension names and coordinates of random variables, instead of specifying the shapes of those random variables as numbers. The difference is whether you intend to change this data after fitting the model, to perform out-of-sample prediciton, for example. 

Reviewer: Maybe I wouldn't start with dims. Perhaps introdução by saying it allows data to be an explicit component of the model, showing in graph_model and being exported to InferenceData after sampling. Then talk about dims and the mutable case?

Author: Sounds better as a motivation for why you'd want to do this. So the "flow" would go:

`ConstantData` --> Look at graphviz/idata --> `ConstantData` with dims --> Look at graphviz/idata --> `MutableData` (with dims) --> etc?

Reviewer: Yeah sounds good

Author: I tried to do something like you suggested, and beefed up some the discussion after each example with some comments on why using Data was helpful in this case (or what it can get you). I hope it's not getting too bloated.

---

(data_container)=
# Using Data Containers

:::{post} Dec 16, 2021
:tags: posterior predictive, shared data 
:category: beginner
:author: Juan Martin Loyola, Kavya Jaiswal, Oriol Abril, Jesse Grabowski
:::

```{code-cell} ipython3
import arviz as az
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import pymc as pm

from numpy.random import default_rng

print(f"Running on PyMC v{pm.__version__}")
```

```{code-cell} ipython3
%matplotlib inline
%config InlineBackend.figure_format = 'retina'
RANDOM_SEED = sum(map(ord, "Data Containers in PyMC"))
rng = default_rng(RANDOM_SEED)
az.style.use("arviz-darkgrid")
```

## Introduction

> After building the statistical model of your dreams, you're going to need to feed it some data. Data is "shown" to a PyMC model via the `observed` keyword argument of all random variables. This argument accepts a wide variety of datatypes, including numpy `ndarrays`, pandas `Series` and `DataFrame`, and even pytensor `TensorVariables`. Although you can pass these "raw" datatypes to your PyMC model, the best way is to use one of two {class}`pymc.Data` containers. 

Reviewer: The other very common case is in `mu = b * x`, so maybe also mention that, or don't focus so much on the `observed` from the beginning

Author: Yes you're right. Now that you're pushing me to do so I can't think of a case where I would want to switch out the observed data.

These containers make it extremely easy to work with data in a PyMC model. They offer a range of benefits, including:

> 1. Visualization of data as a component of your probabilistic graph

Reviewer: Neat!

2. Access to labeled dimensions for readability and accessibility
3. Support for swapping out data for out-of-sample prediction, interpolation/extrapolation, forcasting, etc.

This notebook will illustrate each of these benefits in turn, and show you the best way to integrate data into your PyMC modeling workflow. 

+++

## Types of Data Containers

>  PyMC offers two data containers, depending on your needs: {class}`pymc.ConstantData` and {class}`MutableData`. Both will help you visualize how data fits into your model, store the data in an `InfereceData` for reproducibility, and give access to labeled dimenions. As the names suggest, however, only `MutableData` allows you to change your data to do out-of-sample inference tasks. This ability does, however, come with a small performance cost.

Reviewer: It also allows you to use the same model for different datasets, which can be useful e.g., when doing parameter recovery studies or sensitivity analysis

+++

## Constant Data

The `pm.ConstantData` is a way to add fixed data to a model. It provides a speed boost in exchange for the ability to change the data. If you don't plan on doing out-of-sample prediction, `pm.ConstantData` is for you. 

In the example shows some of the differences between using a data container, in this case `pm.ConstantData`, and "raw" data. This first model shows how raw data, in this case a numpy array, can be directly provided to a PyMC model.

```{code-cell} ipython3
true_mu = 3
true_std = 5
x = rng.normal(loc=true_mu, scale=true_std, size=100)
with pm.Model() as no_data_model:
    mu = pm.Normal("mu")
    sigma = pm.Exponential("sigma", 1)
    obs = pm.Normal("obs", mu=mu, sigma=sigma, observed=x)
```

Looking at the resulting computational graph, the `obs` node is shaded gray to indicate that it has observed data. But the data itself is not shown on the graph.  

```{code-cell} ipython3
pm.model_to_graphviz(no_data_model)
```

In this next model, we create a `pm.ConstantData` container to hold the observations, and pass this container to the `observed`:

```{code-cell} ipython3
with pm.Model() as no_data_model:

>     data = pm.ConstantData("data", x)

Reviewer: Do you think it may be worth to show a simple GLM so we also discuss Data for predictors?

Author: I sort of do that in the section "Using MutableData container variables to predict on new data" -- starting on line 262, using a Logistic regression. It could perhaps be more emphasized that it's a GLM?

Reviewer: I think it's good now

---

(data_container)=
# Using Data Containers

:::{post} Dec 16, 2021
:tags: posterior predictive, shared data 
:category: beginner
:author: Juan Martin Loyola, Kavya Jaiswal, Oriol Abril, Jesse Grabowski
:::

```{code-cell} ipython3
import warnings

import arviz as az
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import pymc as pm

from numpy.random import default_rng

plt.rcParams["figure.constrained_layout.use"] = True

print(f"Running on PyMC v{pm.__version__}")
```

```{code-cell} ipython3
%matplotlib inline
%config InlineBackend.figure_format = 'retina'
RANDOM_SEED = sum(map(ord, "Data Containers in PyMC"))
rng = default_rng(RANDOM_SEED)
az.style.use("arviz-darkgrid")
```

## Introduction

> After building the statistical model of your dreams, you're going to need to feed it some data. Data is typically introduced to a PyMC model in one of two ways. Some data is used as an exogenous input, called `X` in linear regression models, where `mu = X @ beta`. Other data are "observed" examples of the enodgenous outputs of your model, called `y` in regression models, and is used as inputs to the likelihood function implied by your model. These data, either exogenous or endogenous, can be included in your model as wide variety of datatypes, including numpy `ndarrays`, pandas `Series` and `DataFrame`, and even pytensor `TensorVariables`. Although you can pass these "raw" datatypes to your PyMC model, the best way is to use one of two {class}`pymc.Data` containers. 

Reviewer: Typo "enodgenous"

These containers make it extremely easy to work with data in a PyMC model. They offer a range of benefits, including:

1. Visualization of data as a component of your probabilistic graph
2. Access to labeled dimensions for readability and accessibility

> 3. Support for swapping out data for out-of-sample prediction, interpolation/extrapolation, forcasting, etc.

Reviewer: 4. It's stored in the InferenceData after sampling

rng = default_rng(RANDOM_SEED)
az.style.use("arviz-darkgrid")
```

## Introduction

After building the statistical model of your dreams, you're going to need to feed it some data. Data is typically introduced to a PyMC model in one of two ways. Some data is used as an exogenous input, called `X` in linear regression models, where `mu = X @ beta`. Other data are "observed" examples of the endogenous outputs of your model, called `y` in regression models, and is used as input to the likelihood function implied by your model. These data, either exogenous or endogenous, can be included in your model as wide variety of datatypes, including numpy `ndarrays`, pandas `Series` and `DataFrame`, and even pytensor `TensorVariables`. 

Although you can pass these "raw" datatypes to your PyMC model, the best way to introduce data into your model is to use one of two {func}`pymc.Data` containers. These containers make it extremely easy to work with data in a PyMC model. They offer a range of benefits, including:

1. Visualization of data as a component of your probabilistic graph
2. Access to labeled dimensions for readability and accessibility

> 3. Support for swapping out data for out-of-sample prediction, interpolation/extrapolation, forcasting, etc.

Reviewer: ```suggestion
3. Support for swapping out data for out-of-sample prediction, interpolation/extrapolation, forecasting, etc.
```

> 4. All data, not just observed data, will be stored in your {class}`arviz.InferenceData`

Reviewer: ```suggestion
4. All data will be stored in your {class}`arviz.InferenceData`, which is useful for plotting and reproducible workflows.
```

This notebook will illustrate each of these benefits in turn, and show you the best way to integrate data into your PyMC modeling workflow. 

+++

## Types of Data Containers

 PyMC offers two data containers, depending on your needs: {func}`pymc.ConstantData` and {func}`pymc.MutableData`. Both will help you visualize how data fits into your model, store the data in an `InfereceData` for reproducibility, and give access to labeled dimenions. As the names suggest, however, only `MutableData` allows you to change your data. When `X` is `MutableData`, this enables out-of-sample inference tasks. When `y` is `MutableData`, it allows you to reuse the same model on multiple datasets to perform parameter recovery studies or sensitivity analysis.  These abilities do, however, come with a small performance cost.
 
 In past versions of PyMC, the only data container was `pm.Data`. This container is still available for backwards compatability, but the current best practice is to use either `pm.MutableData` or `pm.ConstantData`. 

+++

> ## Constant Data

Reviewer 2: ```suggestion
### Constant Data
```

---

(data_container)=
# Using Data Containers

:::{post} Dec 16, 2021

> :tags: posterior predictive, shared data 

Reviewer 3: and you should also add yourself here in the author metadata. Right now the html hack that added the rendered citation at the end of a page is down, but it will be fixed, and it is not the only nice thing of using that. You will also get a page with the notebooks you have contributed too!

This is not very accessible to not put too much emphasis on authors of pages but you are free to share in on your website, social networks... Here is mine as an example: https://www.pymc.io/projects/examples/en/latest/blog/author/oriol-abril-pla.html

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import pymc as pm

from numpy.random import default_rng

print(f"Running on PyMC v{pm.__version__}")
```

```{code-cell} ipython3
%matplotlib inline
%config InlineBackend.figure_format = 'retina'
RANDOM_SEED = sum(map(ord, "Data Containers in PyMC5"))
rng = default_rng(RANDOM_SEED)
az.style.use("arviz-darkgrid")
```

## Data Containers

The {class}`pymc.Data` container allows users to keep track of dimensions (like dates or cities) and coordinates (such as the actual date times or city names) of multi-dimensional data. It offers two wrappers for this: {class}`pymc.ConstantData` and {class}`MutableData`. Both allow you to specify the dimension names and coordinates of random variables, instead of specifying the shapes of those random variables as numbers. The difference is whether you intend to change this data after fitting the model, to perform out-of-sample prediciton, for example. 

> In past versions of PyMC, the only data container was `pm.Data`. This container is still available for backwards compatability, but the current best practice is to use either `pm.MutableData` or `pm.ConstantData`, depending on your needs.  This notebook will demonstrate how each of these containers are used, and when they are appropriate.

Author: I inferred it from [this warning](https://github.com/pymc-devs/pymc/blob/6bfbff508b80e588cc03711798b4e2577ad46449/pymc/data.py#L432). If you think I'm making the case too strongly I can soften this to something like "you can set `mutable=True/False` in `pm.Data` or be explicit by using `pm.MutableData`/`pm.ConstantData`."

That said in my opinion the recommended best practice should be to use the specific containers instead of the mutable flag,

After building the statistical model of your dreams, you're going to need to feed it some data. Data is typically introduced to a PyMC model in one of two ways. Some data is used as an exogenous input, called `X` in linear regression models, where `mu = X @ beta`. Other data are "observed" examples of the endogenous outputs of your model, called `y` in regression models, and is used as input to the likelihood function implied by your model. These data, either exogenous or endogenous, can be included in your model as wide variety of datatypes, including numpy `ndarrays`, pandas `Series` and `DataFrame`, and even pytensor `TensorVariables`. 

> Although you can pass these "raw" datatypes to your PyMC model, the best way to introduce data into your model is to use one of two {func}`pymc.Data` containers. These containers make it extremely easy to work with data in a PyMC model. They offer a range of benefits, including:

Reviewer 3: That will work. For future reference, the id is the title at the page: https://www.pymc.io/projects/docs/en/latest/api/generated/pymc.Data.html. This does look like a class, but classes are explicited in the docs, for example: https://www.pymc.io/projects/docs/en/latest/api/distributions/generated/pymc.Beta.html.

In the old 3.x days it was different, but now everything is documented at their "recommended import path" as users don't really care that Data is inside the data.py file, only that they can use it as pymc.Data

Author: It would be nice if reference validity was checked as a pre-commit hook

Reviewer 3: This can't be really checked in pre-commit. The docs need to be built to see all the targets that exists and which ones can't be resolved. Ideally, we would have a sphinx build process without warnings so that we could tell readthedocs to fail CI if sphinx prints any warning (like a cross-reference not working). There are currently many warnings (though not as many as in main pymc repo) so we can't do that afaik

model["data"].get_value()
```

Note that we used a pytensor method {meth}`pytensor.compile.sharedvalue.SharedVariable.get_value` of class {class}`pytensor.compile.sharedvalue.SharedVariable` to get the value of the variable. This is because our variable is actually a `SharedVariable`.

```{code-cell} ipython3
type(data)
```

The methods and functions related to the Data container class are:

- `data_container.get_value` (method inherited from the pytensor SharedVariable): gets the value associated with the `data_container`.
- `data_container.set_value` (method inherited from the pytensor SharedVariable): sets the value associated with the `data_container`.
- {func}`pymc.set_data`: PyMC function that sets the value associated with each Data container variable indicated in the dictionary `new_data` with it corresponding new value.

+++

## Using Data container variables to fit the same model to several datasets

> This and the next sections are an adaptation of the notebook ["Advanced usage of pytensor in PyMC"](../Advanced_usage_of_pytensor_in_PyMC.html#using-shared-variables) using `pm.MutableData`.

Reviewer 3: This page no longer exists as it didn't makes sense to keep around. I think we can delete this sentence altogether.

Author: Removed

> az.plot_trace(idata, var_names=["europe_mean_temp", "city_temperature"]);

Author: The auto-placement leaves a bit to be desired -- is there any way to pass `legend_kwargs` to the underlying plot?

```python
    data = pm.ConstantData("data", df_data, dims=("date", "city"))
```

Gave the name "data" to the data, we can access it as follows: 

```{code-cell} ipython3
model["data"].data
```

## MutableData

In many cases, you will want the ability to switch out data between model runs. This arises when you want to fit a model to multiple datasets, or if you are interested in out-of-sample prediction. For these cases, `pm.MutableData` is the correct tool.

+++

### Using MutabelData container variables to fit the same model to several datasets

We can use `MutableData` container variables in PyMC to fit the same model to several datasets without the need to recreate the model each time (which can be time consuming if the number of datasets is large):

```{code-cell} ipython3
# We generate 10 datasets
true_mu = [rng.random() for _ in range(10)]
observed_data = [mu + rng.random(20) for mu in true_mu]

with pm.Model() as model:
    data = pm.MutableData("data", observed_data[0])
    mu = pm.Normal("mu", 0, 10)
    pm.Normal("y", mu=mu, sigma=1, observed=data)
```

> Once again, the name of our data is `data`, so we can look at it's type. Unlike `pm.ConstantData`, we now see a {meth}`pytensor.compile.sharedvalue.SharedVariable.get_value` of class {class}`pytensor.compile.sharedvalue.SharedVariable` to get the value of the variable. This is because our data is now a `SharedVariable`.

Author: It was in the next cell. I split this up and made it more obvious what commentary goes with what output. That said, I think we could just remove this stuff about get_value and set_value, because it's all hidden from users by `pm.set_data`. I've never had to manually interact with the underlying shared variables when working with a pymc model.

The methods and functions related to the Data container class are:

> - `data_container.get_value` (method inherited from the pytensor SharedVariable): gets the value associated with the `data_container`.

Reviewer 3: This can be rendered as a link. Here are multiple options:

```
{meth}`pytensor.compile.sharedvalue.SharedVariable.get_value`  # shows all text, imo, bad idea
{meth}`~pytensor.compile.sharedvalue.SharedVariable.get_value`  # shows only get_value
{meth}`data_container.get_value <pytensor.compile.sharedvalue.SharedVariable.get_value>`  # shows custom text
```

- `data_container.set_value` (method inherited from the pytensor SharedVariable): sets the value associated with the `data_container`.
- {func}`pymc.set_data`: PyMC function that sets the value associated with each Data container variable indicated in the dictionary `new_data` with it corresponding new value.

```{code-cell} ipython3
type(model["data"])
```

To get the values, use the `get_value` method:

```{code-cell} ipython3
model["data"].get_value()
```

In practice, however, you will not need to use `get_value()` or `set_value()`. Instead, it is best to use {func}`pymc.set_data`. Let's use the `MutableData` to repeatedly fit the same model to multiple datasets: 

```{code-cell} ipython3
:tags: [hide-output]

# Generate one trace for each dataset
traces = []
for data_vals in observed_data:
    with model:
        # Switch out the observed dataset
        pm.set_data({"data": data_vals})
        traces.append(pm.sample())
```

> ## Using MutableData container variables to predict on new data

Reviewer 3: this should probably be a level 3 heading, to match with the previous section:

df_data.head()
```

As noted above, `ConstantData` gives you the ability to give named labels to the dimensions of your data. This can be done in two ways:  

> - Entering the dimensions in the `dims` kwarg of a `pm.ConstantData` variable with a pandas Series or DataFrame. The name of the index and columns will be remembered as the dimensions, and PyMC will infer that the values of the given columns must be the coordinates.

Reviewer: I would maybe remove this part. Rather present the basics than the "tricks".

```{code-cell} ipython3
axes = az.plot_trace(idata, var_names=["europe_mean_temp", "expected_city_temp"], legend=True);
```

When we use `pm.ConstantData`, the data are internally represented as a pytensor `TensorConstant`.

```{code-cell} ipython3
type(data)
```

If you need to, you can view the data using the `data` method. All `PyMC` variables, including `Data` containers, can be accessed by indexing the model object with a variable name. Since this line:

```python
    data = pm.ConstantData("observed_temp", df_data, dims=("date", "city"))
```

Gave the name "observed_temp" to the data, we can access it as follows: 

```{code-cell} ipython3
model["observed_temp"].data[:15]
```

## MutableData

In many cases, you will want the ability to switch out data between model runs. This arises when you want to fit a model to multiple datasets, or if you are interested in out-of-sample prediction. For these cases, `pm.MutableData` is the correct tool.

+++

### Using MutableData container variables to fit the same model to several datasets

> We can use `MutableData` container variables in PyMC to fit the same model to several datasets without the need to recreate the model each time (which can be time consuming if the number of datasets is large):

Reviewer: Unfortunately we recompile the d/logp function everything `pm.sample` is called so this is no longer true. It used to be true because we used to cache them

Author: I still think the sentence is true, you're just taking a very strict view of the meaning of "recreate" to mean "recompile". It would still be time consuming to copy/paste a large model over and over to re-use it on multiple datasets.

```{code-cell} ipython3
# We generate 10 datasets
n_models = 10
true_mu = [rng.random() for _ in range(n_models)]
observed_data = [mu + rng.random(20) for mu in true_mu]

with pm.Model() as model:
    data = pm.MutableData("data", observed_data[0])
    mu = pm.Normal("mu", 0, 10)
    pm.Normal("y", mu=mu, sigma=1, observed=data)
```

Once again, the name of our data is `data`, so we can look at it's type. Unlike `pm.ConstantData`, we now see a {meth}`pytensor.compile.sharedvalue.SharedVariable.get_value` of class {class}`pytensor.compile.sharedvalue.SharedVariable` to get the value of the variable. This is because our data is now a `SharedVariable`.

```{code-cell} ipython3
type(model["data"])
```

The methods and functions related to the Data container class are:

- {meth}`data_container.get_value <pytensor.compile.sharedvalue.SharedVariable.get_value>` (method inherited from the pytensor SharedVariable): gets the value associated with the `data_container`.

> - {meth}`data_container.set_value <pytensor.compile.sharedvalue.SharedVariable.set_value>` (method inherited from the pytensor SharedVariable): sets the value associated with the `data_container`.

Reviewer: I don't think it's fruitful to discuss these. Just tell users they can call `.eval` to verify which values are stored in the MutableData?

Author: I like this better (and it's what I always do myself)
